crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -37,26 +31,29 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
         operation: Option<String>,
         options: QueryPlanOptions,
     ) -> Result<Arc<QueryPlan>, QueryPlannerError> {
-        if let Some(value) = self
-            .cached
-            .lock()
-            .await
-            .get(&(query.clone(), operation.clone(), options.clone()))
-            .cloned()
-        {
+        let mut locked_cache = self.cached.lock().await;
+        let key = (query.clone(), operation.clone(), options.clone());
+        if let Some(value) = locked_cache.get(&key).cloned() {
             return value;
         }
 
         let value = self
             .delegate
-            .get(query.clone(), operation.clone(), options.clone())
+            .get(key.0.clone(), key.1.clone(), key.2.clone())
             .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
+        locked_cache.put(key, value.clone());
         value
     }
+
+    async fn get_hot_keys(&self) -> Vec<QueryKey> {
```

Comment:
If the cache is large, we might block all new requests for a long time during the update. Not sure how that can be fixed here, we had RwLocks in various parts of the code before, we can reintroduce one here since writes wot be frequent

Reply:
The work I've done here is to avoid this problem by only waiting on keys which are currently being processed and notifying waiters when the work is complete.